libnmstate/netapplier.py

```diff
@@ -18,20 +18,45 @@
 # Refer to the README and COPYING files for full details of the license
 #
 
+import six
+
+from libnmstate import netinfo
+from libnmstate import nm
 from libnmstate import nmclient
 from libnmstate import validator
 
 
+class UnsupportedIfaceStateError(Exception):
+    pass
+
+
 def apply(desired_state):
     validator.verify(desired_state)
 
-    _apply_ifaces_state(desired_state)
+    current_state = netinfo.interfaces()
+    _apply_ifaces_state(desired_state, current_state)
 
 
-def _apply_ifaces_state(desired_state):
+def _apply_ifaces_state(desired_state, current_state):
     client = nmclient.client()
 
-    for iface_desired_state in desired_state['interfaces']:
+    ifaces_desired_state = _index_by_name(desired_state['interfaces'])
+    ifaces_current_state = _index_by_name(current_state['interfaces'])
+    add_ifaces = [
+        ifaces_desired_state[name] for name in
+        six.viewkeys(ifaces_desired_state) - six.viewkeys(ifaces_current_state)
+    ]
+    edit_ifaces = [
+        ifaces_desired_state[name] for name in
+        six.viewkeys(ifaces_desired_state) & six.viewkeys(ifaces_current_state)
+    ]
+
+    new_con_profiles = [_build_connection_profile(iface_desired_state)
+                        for iface_desired_state in add_ifaces]
+    for connection_profile in new_con_profiles:
+        client.add_connection_async(connection_profile, save_to_disk=True)
```

Comment:
AFAIU this will not do anything as long as there is no GLib.MainLoop running. @thom311 can probably confirm this.

Reply:
Gdbus is a multithreadded D-Bus library, meaning, it uses a separate thread to send/receive D-Bus messages. (I am not sure, but) I think if you just issue asynchronous requests, without ever iterating the main-loop, the requests will still be sent out (at least, if the process lives long enough for the other thread to process the send-queue).

Having said that, I think when asynchronous requests are made, it generally is advised to also listen and handle the response (if only to determine success/failure, and to synchronize with the completion of the request).

Reply:
btw, for many D-Bus requests, NetworkManager needs to authenticate the caller, which generally requires that the process is still around. That means, while you can fire-and-forget style send a D-Bus request and exit (the process), NetworkManager will then be unable to learn the user-id of the process, and fail with permission-denied. For that reason, for many/most D-Bus requests against NetworkManager, you need to keep your process alive long enough. A `sleep` would work, but better of course is to wait for the asynchronous response.

Reply:
It somehow works fine on my machine.
As we discussed already, if we can operate without the mainloop for now, I prefer to postpone handling this after the first PoC due to time restraints.

Reply:
It did not work here with the code already in master that also uses an async method. Addind the MainLoop helped.

Reply:
Do you mean the code from this PR?
Does a sleep works for you?